Implemented K8s patch version masking in the registry matrix of the Installer

jeuxdemains opened the pull request:
**What this PR does / why we need it**:
Implements masking of patch versions of K8s (e.g.: 1.22.*)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #411

**Additional information**
The way it works is the same as the OS version patch (e.g.: Ubuntu_20.04.*). It adds a filter to the patch version witch masks any patch version number.

anusha94 commented on agent/installer/installer.go:
```diff
@@ -52,11 +52,15 @@ func getSupportedRegistry(ob algo.OutputBuilder) registry {
 		// BYOH Bundle Repository. Associate bundle with installer
 		linuxDistro := "Ubuntu_20.04.1_x86-64"
-		addBundleInstaller(linuxDistro, "v1.22.3", &algo.Ubuntu20_4K8s1_22{})
+		k8sVer := "v1.22.3"
```
Why is this hard coding still needed?

anusha94 commented on agent/installer/registry.go:
```diff
@@ -11,18 +11,24 @@ import (
 	osBundle string
 }
-type filterBundleList []filterBundlePair
+type filterK8sBundleMap struct {
```
Suggestion - if we can use only `k8sFilter`, we probably would not need a dummy `v1.22.3` version to be mapped. This makes sense for OS, because we are treating all `20.04.*` as `20.04.1`, but we cannot map the same when it comes to k8s version.

I could be wrong though - I am trying to avoid the hardcoding of `v1.22.3` in `addBundleInstaller` call.

jeuxdemains replied:
Ah, but of course. We don't need a map at all. A simple array with filter(s) will do just fine. Thanks, Anusha!

jeuxdemains replied:
You're totally right. I made the updates accordingly. However, it would be great if you have a look again because I don't feel that I came up with the best solution. I'm open for suggestions.

anusha94 commented on agent/installer/registry_test.go:
```diff
@@ -41,6 +41,9 @@ var _ = Describe("Byohost Installer Tests", func() {
 			Expect(func() { r.AddBundleInstaller("rhel", "1.24", dummy124) }).NotTo(Panic())
 			r.AddOsFilter("ubuntu.*", "ubuntu")
 			r.AddOsFilter("rhel.*", "rhel")
+			r.AddK8sFilter("1.22.*", "1.22")
```
Question - why are we not using complete semver of the k8s version? like `v1.22.3` or `v1.23.5`

jeuxdemains commented:
I removed the hardcoded K8s version in the addBundleInstaller(); Also I refactored the K8s map to an array of filters. Updated the e2e tests as well.

I however have the feeling that my implementation is not good.
The addBundleInstaller() introduces a new argument which is the k8s version -- whenever it is empty it defaults to v1.22.3. I really don't like that and I will need your input for better implementation that will conform the usage of the API by the rest of the TKG components.

I'm very open for comments and opinions.

georgievaVMW commented on agent/installer/installer.go:
```diff
@@ -52,11 +55,14 @@ func getSupportedRegistry(ob algo.OutputBuilder) registry {
 		// BYOH Bundle Repository. Associate bundle with installer
 		linuxDistro := "Ubuntu_20.04.1_x86-64"
-		addBundleInstaller(linuxDistro, "v1.22.3", &algo.Ubuntu20_4K8s1_22{})
+		addBundleInstaller(linuxDistro, k8sVer, &algo.Ubuntu20_4K8s1_22{})
```
I think that the K8s version needs to be specific value and not a variable in this case.
At the moment if well call the method with k8sVer = "1.24" it will again use the bundle installer for "1.22" although different steps might be needed.
As we discussed in our meeting, maybe currently the best option is to keep the version hardcoded but to change from "v1.22.3" to "v1.22"

jeuxdemains replied:
This is something that I overlooked. Alex is right. As we are calling the algorithm for Ubuntu 20.04, Kubernetes 1.22 (algo.Ubuntu20_4K8s1_22), we are bound to hardcode both the OS & K8s major and minor versions. Maybe not directly as a string but it has to always match the algorithm version.
Thank you Alex for your remark.

anusha94 replied:
The algorithm if I understand correctly is just to indicate the sequence of steps. These need not be different for every OS-K8s combination. Is that correct?

Can we do something like -
`addBundleInstaller(linuxDistro, "v1.22.*", &algo.Ubuntu20_4K8s1_22{})`

So that in future it is easier to support other minor versions, like -
`addBundleInstaller(linuxDistro, "v1.23.*", &algo.Ubuntu20_4K8s1_22{})`
Note that the algo is still `1_22` which I think should be fine.

jeuxdemains replied:
Consider it done. Thanks both Anusha and Alex.

anusha94 commented on agent/installer/installer.go:
```diff
@@ -35,8 +35,11 @@ type installer struct {
+func getSupportedRegistry(k8sVer string, ob algo.OutputBuilder) registry {
 	reg := newRegistry()
+	if k8sVer == "" {
+		k8sVer = "v1.22.3"
```
This again is hardcoding to a particular patch version. Can we make this `v1.22.*`?

anusha94 commented on agent/installer/installer.go:
```diff
@@ -52,11 +52,18 @@ func getSupportedRegistry(ob algo.OutputBuilder) registry {
 		linuxDistro := "Ubuntu_20.04.1_x86-64"
-		addBundleInstaller(linuxDistro, "v1.22.3", &algo.Ubuntu20_4K8s1_22{})
+		// If no version specified use the minimum v1.22 supported
+		if k8sVer == "" {
```
We don't have to do this. How about removing `k8sVer` argument from the function itself? Any consequences?

jeuxdemains replied:
No consequences at all, actually I introduced this var in this update in order to apply variable matching. I will update it by reverting some changes now and will do the test. Thanks!

anusha94 commented on agent/installer/installer.go:
```diff
@@ -52,11 +52,18 @@ func getSupportedRegistry(ob algo.OutputBuilder) registry {
+		if k8sVer == "" {
+			k8sVer = "v1.22"
+		}
+		addBundleInstaller(linuxDistro, k8sVer, &algo.Ubuntu20_4K8s1_22{})
```
Here, we can hard code to a particular minor series - 
`addBundleInstaller(linuxDistro, "v1.22.*", &algo.Ubuntu20_4K8s1_22{})`

georgievaVMW reviewed (approved):
LGTM

anusha94 commented on agent/installer/registry.go:
```diff
@@ -11,18 +11,20 @@ import (
 type k8sInstallerMap map[string]osk8sInstaller
 type osk8sInstallerMap map[string]k8sInstallerMap
-type filterBundlePair struct {
+type filterOsBundleMap struct {
```
Why'd you rename this as a map? it isn't a map really.

anusha94 commented on agent/installer/registry_test.go:
```diff
@@ -36,21 +36,24 @@ var _ = Describe("Byohost Installer Tests", func() {
-			Expect(func() { r.AddBundleInstaller("ubuntu", "1.23", dummy123) }).NotTo(Panic())
-			Expect(func() { r.AddBundleInstaller("rhel", "1.24", dummy124) }).NotTo(Panic())
+			Expect(func() { r.AddBundleInstaller("ubuntu", "v1.22.*", dummy122) }).NotTo(Panic())
+			Expect(func() { r.AddBundleInstaller("ubuntu", "v1.23.*", dummy123) }).NotTo(Panic())
```
At this point we don't support `v1.23.*` and `v1.24.*`
These are negative test cases IMO. I understand the assertion is not to panic, but ideally anything other than `v1.22.*` should error out. Do we have such a test case?

jeuxdemains replied:
We don't have test like that, I will convert those to such. Thanks.

anusha94 commented on agent/installer/registry.go:
```diff
@@ -11,18 +11,20 @@ import (
 type registry struct {
 	osk8sInstallerMap
 	filterBundleList
+	k8sFilters []string
```
Also update the comment for the registry struct now that it will include a new field

anusha94 commented on agent/installer/registry.go:
```diff
@@ -11,18 +11,26 @@ import (
+	k8sFilter string
+}
+
+type filterBundleList []filterOsBundlePair
```
```suggestion
type filterOSBundleList []filterOsBundlePair
```

Feel free to include this in one of the future PRs

jeuxdemains replied:
Will do it right away, thanks!

anusha94 reviewed (approved):
lgtm.

anusha94 reviewed (approved):
Thank you for this PR 🎉

dharmjit commented on agent/installer/registry.go:
```diff
@@ -11,18 +11,26 @@ import (
 }
-type filterBundleList []filterBundlePair
+
+type filterK8sBundle struct {
+	k8sFilter string
+}
```
Question: Is there an advantage of defining a struct type to hold only a single string field?
could we instead have?
```
filterk8s string
```

jeuxdemains replied:
It was like that at the beginning but with Anusha we decided to put it into a struct in order to be consistent with the OS bundle struct.

dharmjit reviewed (commented):
lgtm, only a basic nit.
